apps/api/migrations/preference-centralization/preference-centralization.migration.ts

```diff
@@ -0,0 +1,154 @@
+/* eslint-disable no-console */
+import '../../src/config';
+
+import { NestFactory } from '@nestjs/core';
+import { SubscriberPreferenceRepository, NotificationTemplateRepository, PreferenceLevelEnum } from '@novu/dal';
+import {
+  UpsertPreferences,
+  UpsertWorkflowPreferencesCommand,
+  UpsertUserWorkflowPreferencesCommand,
+  UpsertSubscriberGlobalPreferencesCommand,
+  UpsertSubscriberWorkflowPreferencesCommand,
+} from '@novu/application-generic';
+import { buildWorkflowPreferencesFromPreferenceChannels, DEFAULT_WORKFLOW_PREFERENCES } from '@novu/shared';
+
+import { AppModule } from '../../src/app.module';
+
+const BATCH_SIZE = 100;
+
+const sleep = (ms: number) => {
+  return new Promise((resolve) => {
+    setTimeout(resolve, ms);
+  });
+};
+
+/**
+ * Migration to centralize workflow and subscriber preferences.
+ * - subscriber global preference
+ *    -> preferences with subscriber global type
+ * - subscriber workflow preferences
+ *    -> preferences with subscriber workflow type
+ * - workflow preferences
+ *   -> preferences with workflow-resource type
+ *   -> preferences with user-workflow type
+ */
+export async function preferenceCentralization() {
+  const app = await NestFactory.create(AppModule, {
+    logger: false,
+  });
+  console.log('start migration - preference centralization');
+
+  const upsertPreferences = app.get(UpsertPreferences);
+
+  const subscriberPreferenceRepository = app.get(SubscriberPreferenceRepository);
+  const subscriberPreferenceCursor = await subscriberPreferenceRepository._model
+    .find({})
+    .batchSize(BATCH_SIZE)
+    .cursor();
+
+  const counter = {
+    subscriberGlobal: {
+      success: 0,
+      error: 0,
+    },
+    subscriberWorkflow: {
+      success: 0,
+      error: 0,
+    },
+    subscriberUnknown: {
+      success: 0,
+      error: 0,
+    },
+    workflow: {
+      success: 0,
+      error: 0,
+    },
+  };
+  console.log('start workflow preference migration');
+
+  const workflowPreferenceRepository = app.get(NotificationTemplateRepository);
+  const workflowPreferenceCursor = await workflowPreferenceRepository._model.find({}).batchSize(BATCH_SIZE).cursor();
+
+  for await (const workflowPreference of workflowPreferenceCursor) {
+    try {
```

Comment:
Should we add random delay/sleep intervals to give some breathing room to the database?

Reply:
Yes we should - I added a 20% jitter on 10ms between each call.